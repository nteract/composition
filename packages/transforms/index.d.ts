// Type stubs until this package can be converted to TypeScript

export function richestMimetype(bundle: object, order?: any, tf?: any): string;
export const displayOrder: any;
export const transforms: any;
