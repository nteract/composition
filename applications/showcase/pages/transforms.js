// @flow
import * as React from "react";

import {
  richestMimetype,
  standardDisplayOrder,
  standardTransforms
} from "@nteract/transforms";

// Jupyter style MIME bundle
const bundle = {
  "text/plain": "This is great",
  "image/png": "R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7",
  "application/vdom.v1+json": {
    tagName: "h1",
    children: "raw data âž¡ rendered view",
    attributes: {}
  }
};

// Find out which mimetype is the richest
const mimetype = richestMimetype(
  bundle,
  standardDisplayOrder,
  standardTransforms
);

// Get the matching React.Component for that mimetype
let Transform = standardTransforms[mimetype];

export default () => {
  return (
    <div className="root">
      <h1>Showcasing nteract components</h1>

      <hr />

      <h2>
        <pre>@nteract/transforms</pre>
      </h2>
      <Transform data={bundle[mimetype]} />

      <style jsx>{`
        font-family: "Source Sans Pro", Helvetica Neue, Helvetica, Arial,
          sans-serif;
      `}</style>
    </div>
  );
};
