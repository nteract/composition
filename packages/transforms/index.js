/* @flow */
import { List as ImmutableList, Map as ImmutableMap } from 'immutable';

import TextDisplay from './text';
import JsonDisplay from './json';
import JavaScriptDisplay from './javascript';
import HTMLDisplay from './html';
import MarkdownDisplay from './markdown';
import LaTeXDisplay from './latex';

import SVGDisplay from './svg';

import {
  PNGDisplay,
  JPEGDisplay,
  GIFDisplay,
} from './image';

/**
 * Thus begins our path for custom mimetypes and future extensions
 */
import PlotlyTransform from './plotly';
import GeoJSONTransform from './geojson';

import ModelDebug from './model-debug';

import {
  VegaLite,
  Vega,
} from './vega';

type StandardTransforms = ImmutableMap<string, any>
type StandardDisplayOrder = ImmutableList<string>

export const standardTransforms: StandardTransforms = new ImmutableMap({
  'text/plain': TextDisplay,
  'image/png': PNGDisplay,
  'image/jpeg': JPEGDisplay,
  'image/gif': GIFDisplay,
  'image/svg+xml': SVGDisplay,
  'text/html': HTMLDisplay,
  'text/markdown': MarkdownDisplay,
  'text/latex': LaTeXDisplay,
  'application/json': JsonDisplay,
  'application/javascript': JavaScriptDisplay,
});

export const standardDisplayOrder: StandardDisplayOrder = new ImmutableList([
  'application/json',
  'application/javascript',
  'text/html',
  'image/svg+xml',
  'text/markdown',
  'text/latex',
  'image/svg+xml',
  'image/gif',
  'image/png',
  'image/jpeg',
  'application/pdf',
  'text/plain',
]);

function registerTransform({ transforms, displayOrder }, transform) {
  return {
    transforms: transforms.set(transform.MIMETYPE, transform),
    displayOrder: displayOrder.insert(0, transform.MIMETYPE)
  };
}

const additionalTransforms = [
  ModelDebug,
  PlotlyTransform,
  GeoJSONTransform,
  VegaLite,
  Vega,
];

const {
  transforms,
  displayOrder,
} = additionalTransforms.reduce(registerTransform, {
  transforms: standardTransforms,
  displayOrder: standardDisplayOrder,
});

/**
 * Choose the richest mimetype available based on the displayOrder and transforms
 * @param  {ImmutableMap}   bundle - Map({mimetype1: data1, mimetype2: data2, ...})
 * @param  {ImmutableList}  ordered list of mimetypes - List(['text/html', 'text/plain'])
 * @param  {ImmutableMap}   mimetype -> React Component - Map({'text/plain': TextTransform})
 * @return {string}          Richest mimetype
 */

function richestMimetype(bundle: ImmutableMap<string, any>,
  order: ImmutableList<string> = displayOrder,
  tf: ImmutableMap<string, any> = transforms): string {
  return bundle.keySeq()
    // we can only use those we have a transform for
    .filter(mimetype => tf.has(mimetype) && order.includes(mimetype))
    // the richest is based on the order in displayOrder
    .sortBy(mimetype => order.indexOf(mimetype))
    .first();
}

export {
  displayOrder,
  transforms,
  richestMimetype,
};
