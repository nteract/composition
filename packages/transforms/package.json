{
  "name": "@nteract/transforms",
  "version": "5.0.0-alpha.0",
  "description": "Common transforms for Jupyter",
  "main": "lib/index.js",
  "nteractDesktop": "src/index.js",
  "scripts": {
    "prepare": "npm run build",
    "prepublishOnly": "npm run build && npm run build:flow",
    "build": "npm run build:clean && npm run build:lib",
    "build:clean": "rimraf lib",
    "build:flow": "flow-copy-source -v -i '**/__tests__/**' src lib",
    "build:lib": "babel -d lib src --ignore '**/__tests__/**' --config-file ../../babel.config.js",
    "build:lib:watch": "npm run build:lib -- --watch",
    "build:watch": "npm run build:clean && npm run build:lib:watch && npm run build:flow"
  },
  "repository": "https://github.com/nteract/nteract/tree/master/packages/transforms",
  "keywords": [
    "nteract",
    "transforms",
    "notebook"
  ],
  "author": "Kyle Kelley <rgbkrk@gmail.com>",
  "license": "BSD-3-Clause",
  "bugs": {
    "url": "https://github.com/nteract/nteract/issues"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@babel/runtime-corejs2": "^7.0.0",
    "@nteract/markdown": "^3.0.1",
    "@nteract/mathjax": "^3.0.1",
    "@nteract/transform-vdom": "^3.0.0-alpha.0",
    "ansi-to-react": "^4.0.0-alpha.0",
    "babel-runtime": "^6.26.0",
    "react-json-tree": "^0.11.0"
  },
  "peerDependencies": {
    "react": "^16.3.2"
  }
}
