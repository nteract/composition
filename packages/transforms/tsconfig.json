{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src"
  },
  "include": ["src"],
  "references": [
    { "path": "../markdown" },
    { "path": "../mathjax" },
    { "path": "../transform-vdom" }
  ]
}
