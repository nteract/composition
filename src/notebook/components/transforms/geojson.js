/* eslint class-methods-use-this: 0 */

import React, { PropTypes } from 'react';

const L = require('leaflet');

L.Icon.Default.imagePath = '../node_modules/leaflet/dist/images/';

const MIMETYPE = 'application/vnd.geo+json';

export class GeoJSONTransform extends React.Component {


  componentDidMount() {
    this.map = L.map(this.el);
    this.map.scrollWheelZoom.disable();
    // TODO: Determine a strategy for picking tiles
    L.tileLayer('https://api.tiles.mapbox.com/v4/{id}/{z}/{x}/{y}.png?access_token=pk.eyJ1IjoibWFwYm94IiwiYSI6ImNpandmbXliNDBjZWd2M2x6bDk3c2ZtOTkifQ._QA7i5Mpkd_m30IGElHziw', {
      attribution: '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a>',
      id: `mapbox.${this.props.theme}`,
    }).addTo(this.map);
    const geoJSON = this.props.data.toJS();
    const geoJSONLayer = L.geoJson(geoJSON).addTo(this.map);
    const bounds = geoJSONLayer.getBounds();
    this.map.fitBounds(bounds);
  }

  shouldComponentUpdate(nextProps) {
    if (nextProps && nextProps.theme && this.props && nextProps.theme !== this.props.theme) {
      return true;
    }
    return false;
  }

  componentDidUpdate() {
    const theme = this.props.theme === 'light' ||
                    this.props.theme === 'dark' ? this.props.theme : 'light';

    L.tileLayer('https://api.tiles.mapbox.com/v4/{id}/{z}/{x}/{y}.png?access_token=pk.eyJ1IjoibWFwYm94IiwiYSI6ImNpandmbXliNDBjZWd2M2x6bDk3c2ZtOTkifQ._QA7i5Mpkd_m30IGElHziw', {
      attribution: '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a>',
      id: `mapbox.${theme}`,
    }).addTo(this.map);
  }

  render() {
    return (
      <div>
        <link rel="stylesheet" href="../node_modules/leaflet/dist/leaflet.css" />
        <div
          ref={(el) => { this.el = el; }}
          style={{ height: '600px', width: '100%' }}
        />
      </div>
    );
  }
}

GeoJSONTransform.propTypes = {
  data: PropTypes.object,
  theme: PropTypes.string,
};

GeoJSONTransform.defaultProps = {
  theme: 'light',
};

GeoJSONTransform.MIMETYPE = MIMETYPE;

export default GeoJSONTransform;
