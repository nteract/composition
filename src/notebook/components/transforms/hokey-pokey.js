/* eslint class-methods-use-this: 0 */

import React, { PropTypes } from 'react';

const MIMETYPE = 'application/hokey+json';

export class HokeyTransform extends React.Component {
  render() {
    console.warn(this.props);
    const comms = this.props.comms;
    if (!comms) { return null; }

    return (
      <pre>
      {
        JSON.stringify(comms.toJS())
      }
      </pre>
    );
  }
}

HokeyTransform.propTypes = {
  comms: PropTypes.any,
};

HokeyTransform.MIMETYPE = MIMETYPE;

export default HokeyTransform;
