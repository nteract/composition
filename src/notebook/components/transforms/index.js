import { displayOrder, transforms } from 'transformime-react';

import VictoryTransform from './victory';
import PlotlyTransform from './plotly';
import VegaTransform from './vega';

const defaultDisplayOrder = displayOrder
  .splice(0, 0, VegaTransform.MIMETYPE)
  .splice(0, 0, VictoryTransform.MIMETYPE)
  .splice(0, 0, PlotlyTransform.MIMETYPE);

const defaultTransforms = transforms
  .set(VegaTransform.MIMETYPE, VegaTransform)
  .set(VictoryTransform.MIMETYPE, VictoryTransform)
  .set(PlotlyTransform.MIMETYPE, PlotlyTransform);

export { defaultDisplayOrder, defaultTransforms };
