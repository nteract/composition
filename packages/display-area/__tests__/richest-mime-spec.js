import React from "react";

import { shallow } from "enzyme";

import { displayOrder, transforms } from "@nteract/transforms";
import { RichestMime } from "../";
import toJson from "enzyme-to-json";

describe("RichestMime", () => {
  it("renders a mimebundle", () => {
    const models = {};
    const rm = shallow(
      <RichestMime
        displayOrder={displayOrder}
        transforms={transforms}
        bundle={{ "text/plain": "THE DATA" }}
        metadata={{ "text/plain": "alright" }}
        models={models}
      />
    );

    expect(rm.instance().shouldComponentUpdate()).toBeTruthy();

    expect(toJson(rm)).toMatchSnapshot();
  });

  it("does not render unknown mimetypes", () => {
    const rm = shallow(
      <RichestMime
        displayOrder={displayOrder}
        transforms={transforms}
        bundle={{ "application/ipynb+json": "{}" }}
      />
    );

    expect(rm.type()).toBeNull;
  });
});
