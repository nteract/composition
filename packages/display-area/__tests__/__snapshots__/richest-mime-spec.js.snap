// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichestMime renders a mimebundle 1`] = `
<React.Fragment>
  <div>
    <TextDisplay
      data="THE DATA"
      expanded={false}
      metadata="alright"
      models={Object {}}
      theme="light"
    />
  </div>
</React.Fragment>
`;
